Made expandable navbar section more keyboard accessible

alexandrevryghem opened the pull request:
## References
* Fixes #3085

## Description
Improved the accessibility of the expandable navbar menu section. For all the keyboard interactions, I based them on the [W3C guidelines](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/). The only behaviors I haven't implemented are the `Left Arrow` & `Right Arrow`.

## Instructions for Reviewers
List of changes in this PR:
* Created new `HoverOutsideDirective` to detect when you hover outside a specific element to close the expandable modal. This also works when you open it using your keyboard and hover outside the expandable navbar afterwards
* Added multiple new keyboard interaction behaviors:
  * Set focus on the first item when expanding the expandable navbar (with `space` & `enter`)
  * Added `Up Arrow` & `Down Arrow` support to navigate through the expandable navbar options
  * Close the expandable navbar with `esc` & `shift` + `tab`
  * Close the expandable navbar on `tab` and navigate to the next item

The only two keyboard interactions I didn't implement are `Left Arrow` & `Right Arrow`. To achieve these, we would need to know what the next & previous top-level items are. The cleanest approach would have been to trigger a `tab`/`shift` + `tab` event; however, doing this doesn't automatically set the focus on the next/previous top-level item, as this is controlled by the browser. Calculating these items using CSS selectors also didn't seem like a clean solution, which is why I didn't implement this.

**Guidance for how to test and review this PR:**
- Verify that all the keyboard interactions from the [W3C guidelines](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/) now work (except for `Left Arrow` & `Right Arrow`).
- Some screen readers, like [NVDA](https://www.nvaccess.org/download/), block certain keys, causing specific keydown events to not work properly. If you have any screen readers, try enabling them to verify that this still works.

## Checklist
- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue reviewed (changes requested):
@alexandrevryghem : I finally got around to testing/reviewing this today.  Overall this PR *works* and it fixes the dropdown bug (where it stays open if you are using the keyboard).

However, I noticed a very odd display issue **when I'm using my mouse** on the menu.  If I hover over the "All of DSpace" menu with a mouse, **sometimes** it looks like this:
![homepage](https://github.com/user-attachments/assets/35194274-e5e7-43d3-8c81-86e6f51321a2)

I find it odd that the "By Issue Date" option is highlighted by default...this should **only occur** if you are tabbing to it via a keyboard.

This odd display issue doesn't occur *every time*.  But, I've been able to consistently reproduce it if I do this:
* Install this PR
* Run the UI in production mode (`npm run build:prod; npm run serve:ssr`)
* Go to the homepage
* Immediately hover over the "All of DSpace" menu
* If that doesn't work, then reload the page, and hover over the menu again.

This is a minor issue, but it still seems like something we should clean up in this PR.

Finally, I noticed a few specs are missing below.  Beyond that, this PR looks good & works.  If you can find time to get it updated early(ish) next week, then we can still possibly include this in 8.1 / 7.6.3

tdonohue reviewed (changes requested):
@alexandrevryghem : I re-tested this today, and I no longer can get the "All of DSpace" menu to stay open when using keyboard navigation.  It opens briefly, and then closes immediately afterwards.  Maybe removing the focus from the first element is causing this?

1. I'm running in Production mode (`npm run build:prod; npm run serve:ssr`)
2. Using Keyboard only, tab to the "All of DSpace" menu.
3. Click Enter. The menu opens briefly and closes again. If I hold down Enter, then the menu stays open, but I cannot select an option via the keyboard.

This had worked via keyboard navigation previously.  So, recent changes must be the cause.

alexandrevryghem commented:
@tdonohue: Thnx! Indeed it looks like removing that focus broke the functionality for Enter, added a small fix [here](https://github.com/DSpace/dspace-angular/pull/3581/commits/82ed3aadff3203a58087530a483d97fdf62adf2a#diff-cdd3e71164c4bed9d67b431da4766b88683bb5b031dc824ffc475e53dc2c1715R247) and added an additional test to cover this case

tdonohue reviewed (approved):
👍 Thanks @alexandrevryghem !  Retested again, and it's now working properly both via keyboard navigation & mouse navigation.  Code and tests look great too

tdonohue commented:
@alexandrevryghem : This was able to be ported automatically to 8.x, but the port to 7.x failed.  Would you be able to manually port this to `dspace-7_x`?
